### PR TITLE
[Diffusion] Add NVTX markers and cudaProfilerApi step-range profiling

### PR DESCRIPTION
**Pull request by mispa-ms**
## Motivation

Diffusion model inference is opaque to profiling tools out of the box. When using Nsight Systems to investigate performance bottlenecks in denoising pipelines (e.g., attention, scheduler step, per-layer compute), there are no NVTX markers and no way to restrict the capture window to a specific step range without modifying source code.

This PR adds two complementary profiling features for diffusion models:

Layerwise NVTX markers — registers PyTorch forward hooks on transformer modules to emit an NVTX range around each layer's forward pass, making per-layer timing visible in Nsight Systems timelines.
cudaProfilerApi step-range capture — allows profiling to start/stop automatically at a specific global denoising step range (e.g., steps 100–150), so only steady-state inference is captured rather than warmup noise.
Together they enable a workflow like:

nsys profile -c cudaProfilerApi -t cuda,nvtx \
  python -m sglang.cli.main serve --model-path ... \
    --enable-layerwise-nvtx-marker \
    SGLANG_DIFFUSION_PROFILE_STEP_RANGE=100-150

## Modifications

runtime/utils/profiler.py — Adds DiffusionStepProfiler singleton to the existing profiler module:

Tracks cumulative global denoising step count across all requests (warmup excluded)
Calls cudaProfilerStart / cudaProfilerStop when the count enters/exits the configured step range (SGLANG_DIFFUSION_PROFILE_STEP_RANGE=START-END)
Rank-aware: only rank 0 issues profiler calls in multi-GPU (SP/TP) setups
ensure_stopped() for guaranteed cleanup in finally blocks
Request/step-range logging for visibility into which requests overlap the capture window
runtime/pipelines_core/stages/denoising.py

_register_nvtx_hooks(): walks transformer and transformer_2 named modules, skips Identity/Dropout variants and weight-tied duplicates, registers pre/post forward hooks that emit nvtx.range_push/pop with module name and input shapes; hook handles stored in self._nvtx_hook_handles
Denoising loop: wraps denoising_loop, each denoising_step_i, predict_noise_cfg, and scheduler_step with with _nvtx(name): context managers (exception-safe); integrates DiffusionStepProfiler.step() per iteration
runtime/pipelines_core/executors/parallel_executor.py / sync_executor.py

Wraps each pipeline stage execution with with _nvtx(stage_name): (no-op when disabled via nullcontext)
runtime/server_args.py — adds --enable-layerwise-nvtx-marker boolean flag (default False)

multimodal_gen/envs.py — registers SGLANG_DIFFUSION_PROFILE_STEP_RANGE env var

cli/main.py — adds if __name__ == "__main__": main() guard

## Accuracy Tests

This PR does not affect model outputs. No changes to model weights, attention kernels, scheduler logic, or sampling. NVTX hooks are read-only observers; cudaProfilerApi calls have no effect on compute results.

## Benchmarking and Profiling

This feature is the profiling capability — no throughput regression expected since both features are opt-in and disabled by default (enable_layerwise_nvtx_marker=False, no SGLANG_DIFFUSION_PROFILE_STEP_RANGE set).

When enabled, the hook overhead is one dict lookup + string format per module forward pass; negligible relative to GPU kernel latency.

**How to use**
Layerwise NVTX only:

python -m sglang.cli.main serve --model-path <path> --enable-layerwise-nvtx-marker
Capture with Nsight Systems (e.g. nsys profile -t cuda,nvtx -o trace.nsys-rep ...).

Step-range capture (cudaProfilerApi): set SGLANG_DIFFUSION_PROFILE_STEP_RANGE=START-END (e.g. 100-150) so profiling runs only for those global denoising steps. Example:

SGLANG_DIFFUSION_PROFILE_STEP_RANGE=100-150 nsys profile -c cudaProfilerApi -t cuda,nvtx -o trace.nsys-rep \
  python -m sglang.cli.main serve --model-path <path> --enable-layerwise-nvtx-marker

**Nsight Systems trace snapshot**
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/cli/main.py [hunk 1]
```diff
@@ -46,3 +46,7 @@ def main():
         generate(args, extra_argv)
     elif args.subcommand == "version":
         version(args, extra_argv)
+
+
+if __name__ == "__main__":
```
Comment (BBuf) [inline python/sglang/cli/main.py:51]: Why we need this change?

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 2]
```diff
@@ -1,11 +1,32 @@
...
+from typing import Optional, Tuple
 
 import torch
 
+from sglang.multimodal_gen import envs
 from sglang.multimodal_gen.runtime.platforms import current_platform
 from sglang.multimodal_gen.runtime.utils.logging_utils import CYAN, RESET, init_logger
 
+
+def _is_primary_rank() -> bool:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:12]: We should try to avoid using variables or functions starting with an underscore as much as possible.

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 3]
```diff
@@ -1,11 +1,32 @@
...
+    """Check if this is the primary rank (rank 0) for cudaProfiler calls.
+    
+    Only the primary rank should call cudaProfilerStart/Stop to avoid
+    redundant profiling data when using multi-GPU (sequence parallelism, etc.).
+    """
+    try:
+        from sglang.multimodal_gen.runtime.distributed import get_world_group
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:19]: Please move it to the top of the file

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 4]
```diff
@@ -1,11 +1,32 @@
...
+        world_group = get_world_group()
+        if world_group is not None:
+            return world_group.rank == 0
+    except (ImportError, RuntimeError):
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:23]: This try catch is not need?

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 5]
```diff
@@ -18,6 +39,198 @@
...
+                f"Invalid step range: start ({start}) must be less than end ({end})"
+            )
+            return None
+        return (start, end)
+    except ValueError as e:
+        logger.warning(f"Failed to parse step range '{range_str}': {e}")
+        return None
+
+
+class DiffusionStepProfiler:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:66]: This class is too long, can we simplify it?

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 6]
```diff
@@ -18,6 +39,198 @@
...
+    """
+    Global denoising step counter with cudaProfilerApi support.
+
+    This class tracks the global denoising step count across all requests,
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:70]: `DiffusionStepProfiler` is described as a global cross-request step profiler, but `ensure_stopped()` is called at the end of every request. That means a profiling range that spans multiple requests will be cut short, and later requests will not restart profiling because `_profiler_started` is already True?

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 7]
```diff
@@ -18,6 +39,198 @@
...
+            f"cudaProfilerStop at global step {self.global_step_count} "
+            f"(range: {self.profile_range[0]}-{self.profile_range[1]})"
+        )
+        torch.cuda.cudart().cudaProfilerStop()
+        
+        # Exit server after profiling if requested (for BTK automation).
+        # os._exit(0) is used instead of sys.exit(0) because sys.exit raises
+        # SystemExit which is caught by asyncio task machinery and does not
+        # terminate the server process.
+        if envs.SGLANG_DIFFUSION_EXIT_AFTER_PROFILE:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py:186]: `SGLANG_DIFFUSION_EXIT_AFTER_PROFILE` looks unsafe in multi-GPU runs: `_stop_cuda_profiler()` is only executed on rank 0, so only rank 0 calls `os._exit(0)` while other ranks may stay alive and hang in distributed synchronization.

**Review by BBuf — COMMENTED** [review]
Any tests can be added?
